### PR TITLE
Don't copy contiguous bytes on reception

### PR DESCRIPTION
**Pull request by fuzzypixelz**
This uses the slices iterator API of zenoh-c to avoid unecessarily copying bytes into an owned slice, if and only if the bytes is made up of exactly one slice.

### CODE REVIEW

**Comment by fuzzypixelz** [comment]
The aim of this pull request is to fix high latency for relatively large topics (e.g. 1 MB `tagus`) in the single-process iRobot benchmark with `--ipc off`. Results were obtained using the Mont Blanc topology on an **Ubuntu 24.04.1 LTS** machine with an **Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz**.

# dd82e84
| node | topic | size_b | mean_us | min_us | max_us |
|--------|--------|--------|--------|--------|--------|
| ponce | tagus | 1000000 | 3016 | 2512 | 5705 |
| geneva | tagus | 1000000 | 3513 | 2997 | 6457 |
| mandalay | tagus | 1000000 | 2678 | 2214 | 5233 |

# eed223a
| node | topic | size_b | mean_us | min_us | max_us |
|--------|--------|--------|--------|--------|--------|
| ponce | tagus | 1000000 | 1651 | 1363 | 2252 |
| geneva | tagus | 1000000 | 1875 | 1557 | 2563 |
| mandalay | tagus | 1000000 | 1390 | 1151 | 1867 |

# eed223a + 401016c
| node | topic | size_b | mean_us | min_us | max_us |
|--------|--------|--------|--------|--------|--------|
| ponce | tagus | 1000000 | 393 | 257 | 1301 |
| geneva | tagus | 1000000 | 657 | 438 | 1701 |
| mandalay | tagus | 1000000 | 108 | 68 | 650 |

**Comment by clalancette** [comment]
Thanks for opening the PR.

We are just in the process of converting to zenoh-cpp as our API in #327.  I'm assuming a similar idea will work there, but until we merge that, I'm going to mark this as a draft PR, and we can revisit it once we've merged that in.

**Comment by fuzzypixelz** [comment]
> Thanks for opening the PR.
> 
> We are just in the process of converting to zenoh-cpp as our API in #327. I'm assuming a similar idea will work there, but until we merge that, I'm going to mark this as a draft PR, and we can revisit it once we've merged that in.

The bytes slice iterator API exists in zenoh-cpp, however there is no `z_bytes_to_slice` equivalent so we would need to manually allocate a buffer in that case or just use `as_vector`. We could coordinate potential changes to the zenoh-cpp API if necessary.

**Comment by clalancette** [comment]
All right, we've now merged in #327.  Please rebase this and make any necessary changes to adapt to zenoh-cpp, and then we can do a full review of it.

**Comment by fuzzypixelz** [comment]
The following are updated benchmark results after switching from the C to the C++ Zenoh API.

# df6ec26
| node | topic | size_b | mean_us | min_us | max_us |
|--------|--------|--------|--------|--------|--------|
| ponce | tagus | 1000000 | 1117 | 642 | 3399 |
| geneva | tagus | 1000000 | 1433 | 868 | 3775 |
| mandalay | tagus | 1000000 | 755 | 351 | 2807 |

# bf7c354
| node | topic | size_b | mean_us | min_us | max_us |
|--------|--------|--------|--------|--------|--------|
| ponce | tagus | 1000000 | 629 | 402 | 2343 |
| geneva | tagus | 1000000 | 873 | 611 | 2674 |
| mandalay | tagus | 1000000 | 304 | 160 | 1943 |

**Comment by fuzzypixelz** [comment]
> All right, we've now merged in #327. Please rebase this and make any necessary changes to adapt to zenoh-cpp, and then we can do a full review of it.

Should be ready for review now.

File: rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp:47]: Since we are removing the use of `std::vector` from this file, please also remove the `#include <vector>`.

File: rmw_zenoh_cpp/src/detail/rmw_subscription_data.hpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/rmw_subscription_data.hpp:54]: Since we are removing the use of `std::vector` from this file, please also remove the `#include <vector>`.

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/payload.hpp]: Could we use real types here instead of `auto`?  I think it will be easier to read.

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/payload.hpp]: So I'm looking at the definition of `zenoh::Slice` in https://github.com/eclipse-zenoh/zenoh-cpp/blob/e8eca99b4eaff0963e52aefd8405909c1552080d/include/zenoh/api/bytes.hxx#L44-L47 .  It looks to me like this is a shallow copy, and thus we are just copying the `data` pointer and `len` here.  Is this safe to do?  How are we guaranteeing that the data that is pointed to by the `data` pointer still exists later on when we try to access it?

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/payload.hpp]:
```suggestion
#include <zenoh.hxx>
```

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/payload.hpp]:
```suggestion
#include <utility>
#include <variant>
#include <vector>
```

**Review by clalancette — CHANGES_REQUESTED** [review]
Besides what I put inline, I'd also prefer not to have the implementation in the header file.  That means that any downstream consumer has to compile the code all of the time.  My suggestion is to just put this whole thing in the `rmw_subscription_data.cpp` file, but if you want to keep it a separate file then I'd suggest keeping just the API in the .hpp file and putting the implementation in a .cpp file.

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (fuzzypixelz) [reply rmw_zenoh_cpp/src/detail/payload.hpp]: That is precisely why we keep a clone of `bytes`, ensuring that the `Bytes` reference count maintained by Zenoh wouldn't go to zero and invalidate the slice.

Comment (clalancette) [reply rmw_zenoh_cpp/src/detail/payload.hpp]: Right, but what's the relationship between the pointer returned by `bytes.clone()`, and `slice.data`?  It looks to me like in the `Contiguous` case, we clone the bytes, but then return the data from the `slice.data` later anyway.

Comment (fuzzypixelz) [reply rmw_zenoh_cpp/src/detail/payload.hpp]: These are roughly the definitions on the Rust side:

```rust
pub struct ZBytes(ZBuf);

pub struct ZBuf {
    slices: SingleOrVec<ZSlice>,
}

pub struct ZSlice {
    buf: Arc<dyn ZSliceBuffer>,
    start: usize,
    end: usize,
    #[cfg(feature = "shared-memory")]
    pub kind: ZSliceKind,
}
```

So `Bytes` is roughly a list of reference-counted buffers. We only need to maintain a non-zero the reference count of the `Arc` to ensure that `slice.data` stays valid, as it points directly to the buffer behind the `Arc`.

See:
- https://github.com/eclipse-zenoh/zenoh/blob/a6bc509b5997c922a56253455149858af715edfa/commons/zenoh-buffers/src/zslice.rs#L92C1-L98C2
- https://github.com/eclipse-zenoh/zenoh/blob/a6bc509b5997c922a56253455149858af715edfa/commons/zenoh-buffers/src/zbuf.rs#L32-L34
- https://github.com/eclipse-zenoh/zenoh/blob/a6bc509b5997c922a56253455149858af715edfa/zenoh/src/api/bytes.rs#L93

**Comment by fuzzypixelz** [comment]
> Besides what I put inline, I'd also prefer not to have the implementation in the header file. That means that any downstream consumer has to compile the code all of the time. My suggestion is to just put this whole thing in the `rmw_subscription_data.cpp` file, but if you want to keep it a separate file then I'd suggest keeping just the API in the .hpp file and putting the implementation in a .cpp file.

I will keep a separate translation unit for `Payload` as it might be useful for queries as well.

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (clalancette) [reply rmw_zenoh_cpp/src/detail/payload.hpp]: Thanks for the explanation.

From what I can tell of your explanation, then, we need to hold onto the `zenoh::Bytes` object as the reference count, and then that keeps the `zenoh::Slice` buffer around until we drop the reference later.  Do I have that right?

Assuming that is the case, I'm going to suggest adding a comment to that effect, as it is very non-obvious in this code.  It might even be worth renaming the `bytes` member of the structure to `ref` as well, but I'll leave that up to you.

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (Yadunund) [inline rmw_zenoh_cpp/src/detail/payload.hpp]: Can we move this definition to the existing `detail/zenoh_utils.hpp` header and have the implementation defined in `detail/zenoh_utils.cpp`?

**Review by Yadunund — CHANGES_REQUESTED** [review]

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (fuzzypixelz) [reply rmw_zenoh_cpp/src/detail/payload.hpp]: See https://github.com/ros2/rmw_zenoh/pull/343#issuecomment-2551797597.

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (fuzzypixelz) [reply rmw_zenoh_cpp/src/detail/payload.hpp]: Please see if the comment I added in e76073831c078012343e315cf3564be68fa02489 is satisfactory.

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/payload.hpp]: include `<cstdint>`

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/payload.hpp]: include `<optional>`

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/payload.hpp]: include `<tuple>`

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (fuzzypixelz) [reply rmw_zenoh_cpp/src/detail/payload.hpp]: [This is `std::get(std::variant)` and is defined in `variant`.](https://en.cppreference.com/w/cpp/utility/variant/get)

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (fuzzypixelz) [reply rmw_zenoh_cpp/src/detail/payload.hpp]: Nevermind.

File: rmw_zenoh_cpp/src/detail/payload.hpp
Comment (clalancette) [reply rmw_zenoh_cpp/src/detail/payload.hpp]: That comment looks great, though I think it got lost in the recent refactor.

**Comment by YuanYuYuan** [comment]
A little suggestion: The `Payload` name seems to be too generic and we don't explain its behavior (which is a copy-on-write when making the bytes contiguous). Can we give it a proper name like `CowPayload` or `CowBytes`?

**Comment by fuzzypixelz** [comment]
79fc5ce304fa4ef9a40ab497e71d8a12d91a2449 seems to have broken the Jazzy build.

**Review by clalancette — APPROVED** [review]
I'm happy with this now.  Thanks for the revisions, I'm going to go ahead and merge this.

**Comment by Yadunund** [comment]
@Mergifyio backport humble jazzy

**Pull request merged**
